### PR TITLE
[WIP, ENH] Add GODEC

### LINKED ISSUE #79: Add GODEC to package
This was initially brought up in the me-ica repository ([here](https://github.com/ME-ICA/me-ica/issues/4)), but after discussing it with @emdupre we figured it would be good to open here as well. This is a low-priority enhancement, but is probably worth incorporating in the long term.

### PR DESCRIPTION
**Pull request by tsalo**
Closes #79. Ref #106.

@dowdlelt I know that you were able to get the GODEC implementation [here](https://bitbucket.org/prantikk/me-ica/src/e21c93a6358ec43eb4c915f2ae42711014e6f3d3/meica.libs/godec.py?at=v3&fileviewer=file-view-default) working. I used that code as well, with several changes to remove redundancy and unused functions. Perhaps you could take a look at this PR?

Changes proposed in this pull request:

- Add Go Decomposition to package under `tedana.decomposition.godecomp.tedgodec`.
- Integrate GODEC into tedana workflow with `ws_denoise` argument. Still haven't added a command-line argument.
- ~~Fix potential bug in MEDN construction within `tedana.utils.io.gscontrol_mmix`.~~ I've removed this to prevent bloat in the PR. I'm going to open a separate PR (#125) for this potential bug fix.
- ~~Replace `label` with `out_dir` and add `out_dir` argument to many functions.~~ Removed to reduce bloat.

To do:
- [ ] Code review
- [ ] Add unit tests
- [x] Update documentation
- [x] Update CLI

### CODE REVIEW

**Comment by dowdlelt** [comment]
I looked for anything obvious, and it seems fine by me. I believe 'vn' is for variance normalization, rather than demeaning. 

I am not sure there I ever actually got the python implementation working correctly, though I do recall getting close. When I brought this up on the MEICA repository I was pointed towards the T1c outputs provided by MEICA, by @prantikk [here](https://github.com/ME-ICA/me-ica/issues/4#issuecomment-369058536). Limited testing suggested that the T1c mimicked the characteristics of the godec output from the 2018 Power et al [paper](http://www.pnas.org/content/115/9/E2105), all with less fiddling. I believe the biggest problem I ran into was in the output from the algorithm, rather than running it. 

In any case, once this is tested it would be nice to compare the T1c data with the GoDec output to confirm (or not) similarities in a single dataset.

**Comment by tsalo** [comment]
Thanks for taking a look @dowdlelt. Also thanks for catching vn's meaning!

Do you know if there's any kind of citation for the minimum image regression? The Power paper (as you know) recommended GODEC or a more standard GSR (which doesn't seem to be implemented in the same way in `gscontrol_raw` in tedana), but I'm guessing that the minimum image regression came later.

Certainly a quantitative comparison of the recommended method from the literature (GODEC) against the new, simpler method recommended here (T1c) would be nice for anyone who wants to publish using T1c.

**Comment by tsalo** [comment]
As expected, the proposed changes break the integration test (specifically for `dn_ts_OC_T1c.nii`).

File: tedana/utils/io.py [hunk 1]
```diff
@@ -54,7 +56,9 @@ def gscontrol_mmix(optcom_ts, mmix, mask, acc, ref_img):
     """
     data_norm = (optcom_masked - optcom_mu) / optcom_std
     cbetas = lstsq(mmix, data_norm.T, rcond=None)[0].T
-    resid = data_norm - np.dot(cbetas, mmix.T)
+    all_comps = np.arange(mmix.shape[0])
+    not_ign = sorted(np.setdiff1d(all_comps, ign))
+    resid = data_norm - np.dot(cbetas[:, not_ign], mmix[:, not_ign].T)
```
[the remark this answers is not in the document]
Comment (tsalo) [reply tedana/utils/io.py]: Moved to new PR.

File: tedana/utils/io.py [hunk 2]
```diff
@@ -54,7 +57,9 @@ def gscontrol_mmix(optcom_ts, mmix, mask, acc, ref_img):
     """
     data_norm = (optcom_masked - optcom_mu) / optcom_std
     cbetas = lstsq(mmix, data_norm.T, rcond=None)[0].T
-    resid = data_norm - np.dot(cbetas, mmix.T)
```
[the remark this answers is not in the document]
Comment (emdupre) [reply tedana/utils/io.py]: Sorry, is this resolved or unresolved ?

File: tedana/utils/io.py [hunk 3]
```diff
@@ -75,14 +79,14 @@ def gscontrol_mmix(optcom_ts, mmix, mask, acc, ref_img):
     """
     bold_noT1gs = bold_ts - np.dot(lstsq(glob_sig.T, bold_ts.T,
                                          rcond=None)[0].T, glob_sig)
-    utils.filewrite(utils.unmask(bold_noT1gs * optcom_std, mask),
-                    'hik_ts_OC_T1c.nii', ref_img)
+    hik_ts = optcom_mu + (bold_noT1gs * optcom_std)
```
[the remark this answers is not in the document]
Comment (emdupre) [reply tedana/utils/io.py]: Yes, the High K time series does not include the mean -- Thanks for confirming !

File: tedana/workflows/tedana.py [hunk 4]
```diff
@@ -201,6 +202,8 @@ def tedana_workflow(data, tes, mask=None, mixm=None, ctab=None, manacc=None,
         Ignore low-variance ambiguous components. Default is False.
     gscontrol : :obj:`bool`, optional
         Control global signal using spatial approach. Default is True.
+    ws_denoise : {None, 'gsr', 'godec'}, optional
+        Which method to apply for widespread signal denoising. Default is None.
```
[the remark this answers is not in the document]
Comment (emdupre) [reply tedana/workflows/tedana.py]: I'm ok with `ws_denoise` ! @dowdlelt, WDYT ?

File: tedana/utils/io.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (tsalo) [reply tedana/utils/io.py]: Yes, this was moved to #125.

File: tedana/workflows/tedana.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (tsalo) [reply tedana/workflows/tedana.py]: This was handled in #163.

File: tedana/workflows/tedana.py [hunk 5]
```diff
@@ -384,17 +374,19 @@ def tedana_workflow(data, tes, mask=None, mixm=None, ctab=None, manacc=None,
     utils.filewrite(s0G, op.join(out_dir, 's0vG.nii'), ref_img)
 
     # optimally combine data
-    OCcatd = model.make_optcom(catd, tes, mask, t2s=t2sG, combmode=combmode)
+    optcom_ts = model.make_optcom(catd, tes, mask, t2s=t2sG, combmode=combmode)
 
     # regress out global signal unless explicitly not desired
     if gscontrol:
```
[the remark this answers is not in the document]
Comment (tsalo) [reply tedana/workflows/tedana.py]: This was handled in #163.

**Comment by jbteves** [comment]
I remembered this PR while re-reading #79. With all of the merges there are no some conflicts. Would you mind updating this branch to be up-to-date with master?

**Pull request closed without merging**